### PR TITLE
Minor revamp of input

### PR DESCRIPTION
**Pull request by rictic**
* Associate input with a player rather than with the game itself.
* Keep track of what raw keyboard buttons are held down. Calculate player input based on the held buttons.
* Support WASD movement
* Only preventDefault on known buttons (so f5 and ctrl-r to refresh work)
* Deal with fresh inputs separately from simulation, so that we handle cases where someone presses a button with an instantaneous effect (like Q) for less than a full frame

### CODE REVIEW

File: www/index.js [hunk 1]
```diff
@@ -72,14 +20,7 @@ function drawOneFrame() {
...
   previousFrameTime = timestamp;
-  const simError = game.simulate(
-    elapsed,
-    curInput.up,
-    curInput.down,
-    curInput.left,
-    curInput.right,
-    curInput.q
-  );
+  const simError = game.simulate(elapsed);
```
Comment (cdata) [inline www/index.js:23]: ✨

File: www/index.js [hunk 2]
```diff
@@ -122,4 +65,31 @@ function average(arr) {
...
+const knownButtons = new Set(
+  ['w', 'a', 's', 'd', 'q', 'arrowup', 'arrowdown', 'arrowleft', 'arrowright']);
+const heldButtons = {};
+function updateInput() {
+  const up = heldButtons['w'] || heldButtons['arrowup'];
+  const down = heldButtons['s'] || heldButtons['arrowdown'];
+  const left = heldButtons['a'] || heldButtons['arrowleft'];
+  const right = heldButtons['d'] || heldButtons['arrowright'];
+  const q = heldButtons['q'];
+  game.set_inputs(up, down, left, right, q);
```
Comment (cdata) [inline www/index.js:77]: 👍

File: src/lib.rs [hunk 3]
```diff
@@ -219,29 +220,36 @@ impl Game {
...
-        &mut self,
-        elapsed: f64,
-        up: bool,
-        down: bool,
-        left: bool,
-        right: bool,
-        q: bool,
-    ) -> Option<String> {
-        let next_input = InputState {
+    fn local_player(&mut self) -> Option<&mut Player> {
```
Comment (cdata) [inline src/lib.rs:223]: Is this better than holding a reference to the local player directly?

**Review by cdata — APPROVED** [review]
Very nice cleanup!

File: src/lib.rs (hunk 3 above)
Comment (rictic) [reply src/lib.rs:223]: Like, adding an `&Player` field on `Game`? That'd be super convenient, but I _think_ that the borrow checker wouldn't like it? Something to experiment with!

**Pull request merged**